stan/math/rev/core/profiling.hpp

```diff
@@ -0,0 +1,167 @@
+      chain_stack_size_sum_ += (ChainableStack::instance_->var_stack_.size()
+                                - start_chain_stack_size_ - 1);
+      nochain_stack_size_sum_
+          += (ChainableStack::instance_->var_nochain_stack_.size()
+              - start_nochain_stack_size_);
+    } else {
+      n_fwd_no_AD_passes_++;
+    }
+    fwd_pass_time_ += std::chrono::duration<double>(
+                          std::chrono::steady_clock::now() - fwd_pass_tp_)
+                          .count();
+    active_ = false;
+  }
+
+  void rev_pass_start() { rev_pass_tp_ = std::chrono::steady_clock::now(); }
+
+  void rev_pass_stop() {
+    rev_pass_time_ += std::chrono::duration<double>(
+                          std::chrono::steady_clock::now() - rev_pass_tp_)
+                          .count();
+    n_rev_passes_++;
+  }
+
+  size_t get_chain_stack_used() const noexcept {
+    return chain_stack_size_sum_;
+  };
+
+  size_t get_nochain_stack_used() const noexcept {
+    return nochain_stack_size_sum_;
+  };
+
+  size_t get_num_no_AD_fwd_passes() const noexcept {
+    return n_fwd_no_AD_passes_;
+  }
+
+  size_t get_num_AD_fwd_passes() const noexcept { return n_fwd_AD_passes_; }
+
+  size_t get_num_fwd_passes() const noexcept {
+    return n_fwd_AD_passes_ + n_fwd_no_AD_passes_;
+  }
+
+  double get_fwd_time() const noexcept { return fwd_pass_time_; }
+
+  size_t get_num_rev_passes() const noexcept { return n_rev_passes_; }
+
+  double get_rev_time() const noexcept { return rev_pass_time_; }
+};
+
+using profile_key = std::pair<std::string, std::thread::id>;
+
+using profile_map = std::map<profile_key, profile_info>;
+
+/**
+ * Profiles C++ lines where the object is in scope.
+ * When T is var, the constructor starts the profile for the forward pass
+ * and places a var with a callback to stop the reverse pass on the
+ * AD tape. The destructor stops the profile for the forward pass and
+ * places a var with a callback to start the profile for the reverse pass.
+ * When T is not var, the constructor and destructor only profile the
+ *
+ *
+ * @tparam T type of profile class. If var, the created object is used
+ * to profile reverse mode AD. Only profiles the forward pass otherwise.
+ */
+template <typename T>
+class profile {
+  profile_key key_;
+  profile_info* profile_;
+
+ public:
+  profile(std::string name, profile_map& profiles)
+      : key_({name, std::this_thread::get_id()}) {
+    profile_map::iterator p = profiles.find(key_);
+    if (p == profiles.end()) {
+      profiles[key_] = profile_info();
+    }
+    profile_ = &profiles[key_];
+    if (profile_->is_active()) {
+      std::ostringstream msg;
+      msg << "Profile '" << key_.first << "' already started!";
+      throw std::runtime_error(msg.str());
+    }
+    profile_->fwd_pass_start<T>();
+    if (!is_constant<T>::value) {
+      profile_info& profile_ref = *profile_;
+      reverse_pass_callback([&profile_ref]() mutable { profile_ref.rev_pass_stop(); });
```

Comment:
I need to do 
```cpp
profile_info& profile_ref = *profile_;
```
in order for this to pass the address sanitizer. Is there a nicer way?

cc @t4c1

Reply:
```
reverse_pass_callback([profile_]() mutable { profile_->rev_pass_stop(); });
```

I think. Copy the pointer by value and use it in the reverse pass.

Reply:
> reverse_pass_callback([profile_]() mutable { profile_->rev_pass_stop(); });

Does not compile.

> reverse_pass_callback([this]() mutable { profile_->rev_pass_stop(); });

compiles but fails withe the address sanitizer.

Reply:
I am fine with the way its done now as it looks to be safe and it also works. But if there is a shorter way I am all ears.

Reply:
You can try:
`reverse_pass_callback([profile=this->profile_]() mutable { profile->rev_pass_stop(); })`
But I have no idea if it will work.

Reply:
Works! Tadej to the rescue once again. Thanks.